lib/climate_control.rb

```diff
@@ -1,16 +1,44 @@
-require "climate_control/environment"
 require "climate_control/errors"
-require "climate_control/modifier"
 require "climate_control/version"
+require "monitor"
 
 module ClimateControl
-  @@env = ClimateControl::Environment.new
+  extend self
 
-  def self.modify(environment_overrides = {}, &block)
-    Modifier.new(env, environment_overrides, &block).process
+  SEMAPHORE = Monitor.new
+  private_constant :SEMAPHORE
+
+  def modify(environment_overrides = {}, &block)
+    environment_overrides = environment_overrides.transform_keys(&:to_s)
+
+    SEMAPHORE.synchronize do
+      previous = ENV.to_hash
+      middle = {}
+      copy environment_overrides
+      middle = ENV.to_hash
+      block.call
+    ensure
+      after = ENV
+      (previous.keys | middle.keys | after.keys).each do |key|
+        if previous[key] != after[key] && middle[key] == after[key]
+          ENV[key] = previous[key]
+        end
+      end
+    end
   end
 
-  def self.env
-    @@env
+  def env
+    ENV
+  end
```

Comment:
🤔 I like this method, but b/c we directly reference the constant `ENV` up above, this method can be incorrect, and thus can't really be depended on. Eg if something mixed this module in and overrode this method, then it would not behave as expected.

IMO, this method is fine, and the other code should reference `ENV` by calling this method. If we don't want to do that, then it feels like we should remove this method since it can't really be correct.

Reply:
I'm thinking of depreciating this method, what do you think?

Something like:

```ruby
warn "ClimateControl.env is depreciated. ENV should be used instead"
```

Reply:
Opened an issue to discuss #44